interfaces/builtin/overmount.go

```diff
@@ -0,0 +1,136 @@
+	return "overmount"
+}
+
+func (iface *OvermountInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	return nil
+}
+
+func (iface *OvermountInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
+	}
+
+	// check that we both a source and a destination
+	source, destination, err := iface.getAttribs(plug.Attrs)
+	if err != nil {
+		return err
+	}
+	if len(source) == 0 || len(destination) == 0 {
+		return fmt.Errorf("source and destination must be set")
+	}
+
+	return nil
+}
+
+func (iface *OvermountInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+func (iface *OvermountInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+func (iface *OvermountInterface) overmountMountSnippet(plug *interfaces.Plug, src, dst string) ([]byte, error) {
+	snippet := bytes.NewBuffer(nil)
+	fmt.Fprintln(snippet, overmountEntry(plug, src, dst, ",ro,exec"))
+	return snippet.Bytes(), nil
+}
+
+// Obtain yaml-specified source and destination to mount
+func (iface *OvermountInterface) getAttribs(attribs map[string]interface{}) (string, string, error) {
+	// source: path within the snap
+	source, ok := attribs["source"].(string)
+	if !ok {
+		return "", "", fmt.Errorf("cannot find attribute 'source'")
+	}
+
+	// destination: absolute path to mount the source
+	destination, ok := attribs["destination"].(string)
+	if !ok {
+		return "", "", fmt.Errorf("cannot find attribute 'destination'")
+	}
+
+	return source, destination, nil
+}
+
+func appArmorEntry(plug *interfaces.Plug, src, dst, mntOpts, permissions string) string {
+	return fmt.Sprintf("mount options=(ro bind %s) %s/** -> %s/**,\n%s/** %s,", mntOpts, src, dst, dst, permissions)
+}
+
+func overmountEntry(plug *interfaces.Plug, src, dst string, mntOpts string) string {
+	return fmt.Sprintf("%s %s none bind%s 0 0", src, dst, mntOpts)
+}
+
+func (iface *OvermountInterface) overmountAppArmorSnippet(plug *interfaces.Plug, src, dst string) ([]byte, error) {
+	snippet := bytes.NewBuffer(nil)
+	fmt.Fprintln(snippet, appArmorEntry(plug, src, dst, "exec", "mrkix"))
+	return snippet.Bytes(), nil
+}
+
+func (iface *OvermountInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	source, destination, err := iface.getAttribs(plug.Attrs)
+	if err != nil {
+		return nil, err
+	}
+
+	src := resolveSpecialVariable(source, plug.Snap)
+	if !strings.HasPrefix(destination, "/") {
+		return nil, fmt.Errorf("destination must be an absolute path: %q", destination)
+	}
+	if strings.HasPrefix(destination, "/snap/") || strings.HasPrefix(destination, "/var/snap/") {
```

Comment:
The checks here are good, but they should be done on sanitize rather than waiting for connect. Or even better, in the common `mounts()` function.

Let's please also ensure source is prefixed by `$SNAP/` or `$SNAP_DATA/` and let's also make sure neither `source` nor `target` contain "..". Both should be done before resolution, so we're validating the original path used instead of what we already processed.